File: leetcode/34-find-first-and-last-position-of-element-in-sorted-array.py
# https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/ (top-interview-questions)

from typing import List


class Solution:
	def searchRange(self, nums: List[int], target: int) -> List[int]:
		# initiate the starting and the ending position of the target
		start, end = -1, -1

		# find the starting position of the target
		left, right = 0, len(nums)-1
		while left <= right:
			mid = left + (right - left) // 2
			if nums[mid] < target: left = mid + 1
			elif nums[mid] > target: right = mid - 1
			else: # in case we found him we just save this index value and we continue our binary search in the left part of the array
				start = mid
				right = mid - 1

		# find the ending position of the target
		left, right = 0, len(nums)-1
		while left <= right:
			mid = left + (right - left) // 2
			if nums[mid] < target: left = mid + 1
			elif nums[mid] > target: right = mid - 1
			else: # in case we found him we just save this index value and we continue our binary search in the right part of the array
				end = mid
				left = mid + 1

		return [start, end]

File: leetcode/34-find-first-and-last-position-of-element-in-sorted-array.js
// https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/ (top-interview-questions)

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function(nums, target) {
  let starting = -1;
  let ending = -1;

  let l = 0;
  let m;
  let r = nums.length - 1;

  while (l<=r) {
    m = l + Math.floor((r-l) / 2);
    if (target > nums[m]) {
      l = m + 1;
    } else if (target < nums[m]) {
      r = m - 1;
    } else {
      starting = m;
      r = m - 1;
    }
  }

  l = 0;
  r = nums.length - 1;
  while (l<=r) {
    m = l + Math.floor((r-l) / 2);
    if (target > nums[m]) {
      l = m + 1;
    } else if (target < nums[m]) {
      r = m - 1;
    } else {
      ending = m;
      l = m + 1;
    }
  }

  return [starting, ending];
};



searchRange([5,7,7,8,8,10], 8);